### PR TITLE
GSW-1145 feat: distribution calculation in emission

### PR DESCRIPTION
**Pull request by r3v4s**
## feat
1. distribution logic
- transfer uGNS to each target with distribution percent
- each target distribution percent can be change by admin or governance.

### CODE REVIEW

**Comment by jungju** [comment]
In Golang, comments for functions and types should start with their respective names. The order of code declarations in Golang is import, const, type, var, and func. Does this also apply to Gno?

**Comment by notJoon** [comment]
> In Golang, comments for functions and types should start with their respective names. The order of code declarations in Golang is import, const, type, var, and func. Does this also apply to Gno?

Yes, actually gno also encourages to follow the godoc style conventions.

**Comment by r3v4s** [comment]
> > In Golang, comments for functions and types should start with their respective names. The order of code declarations in Golang is import, const, type, var, and func. Does this also apply to Gno?
> 
> Yes, actually gno also encourages to follow the godoc style conventions.


Following below sequence
```
interface - type - const - var - new func - public receiver func - private receiver func - public func - private func
```

File: emission/distribution.gno [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import (
+	"std"
+
+	"gno.land/r/demo/gnoswap/consts"
+	"gno.land/r/demo/gns"
+
+	"gno.land/p/demo/ufmt"
+)
+
+type emissionTarget int
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno:13]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	COMMUNITY_POOL
+	XGNS
+)
+
+// BasisPoint
+// 1 = 0.01%
+// 10 = 0.1%
+// 100 = 1%
+// 1000 = 10%
+// 10000 = 100%
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+type distributionPctMap map[emissionTarget]uint64
+
+var distributionPct distributionPctMap = distributionPctMap{
+	LIQUIDITY_STAKING: 7500, // / 10000 = 75%
+	DEVOPS:            2000,
+	COMMUNITY_POOL:    500,
+	XGNS:              0,
+}
+
+func DistributeToTarget(amount uint64) {
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno:34]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		totalSent += distAmount
+
+		transferToTarget(target, distAmount)
+	}
+
+	// `amount-totalSent` can be left due to rounding
+	// it will be distributed next time
+}
+
+func getAmount(amount, bptPct uint64) uint64 {
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		distAmount := getAmount(amount, pct)
+		totalSent += distAmount
+
+		transferToTarget(target, distAmount)
+	}
+
+	// `amount-totalSent` can be left due to rounding
+	// it will be distributed next time
+}
+
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno:46]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	changeDistributionPct(target04, pct04)
+
+	checkSumDistributionPct()
+}
+
+func changeDistributionPct(target emissionTarget, pct uint64) {
+	// only admin or governance can change
+	caller := std.PrevRealm().Addr()
+	if caller != consts.GNOSWAP_ADMIN && caller != consts.GOV_ADDR {
+		panic("only admin or governance can change")
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+	// `amount-totalSent` can be left due to rounding
+	// it will be distributed next time
+}
+
+func getAmount(amount, bptPct uint64) uint64 {
+	return amount * bptPct / 10000
+}
+
+func transferToTarget(target emissionTarget, amount uint64) {
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	if caller != consts.GNOSWAP_ADMIN && caller != consts.GOV_ADDR {
+		panic("only admin or governance can change")
+	}
+
+	// cannot add new target
+	if target != LIQUIDITY_STAKING && target != DEVOPS && target != COMMUNITY_POOL && target != XGNS {
+		panic("invalid target")
+	}
+
+	// maximum pct for single target is 100% (10000 bps)
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	if pct > 10000 {
+		panic("too many pct")
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	distributionPct[target] = pct
+}
+
+// after changing distribution pct, check if sum of all pct is 100% (10000 bps)
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno]: 0de2ab4420588b76435b6cba2fae973efaec8170

File: emission/distribution.gno [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+// BasisPoint
+// 1 = 0.01%
+// 10 = 0.1%
+// 100 = 1%
+// 1000 = 10%
+// 10000 = 100%
+type distributionPctMap map[emissionTarget]uint64
+
+var distributionPct distributionPctMap = distributionPctMap{
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply emission/distribution.gno:26]: 0de2ab4420588b76435b6cba2fae973efaec8170

**Review by dongwon8247 — DISMISSED** [review]

File: emission/__TEST_0_INIT_VARIABLE_AND_HELPER_test.gno [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+// Realms to mock frames
+var (
+	gsaRealm = std.NewUserRealm(gsa)
+
+	govRealm = std.NewCodeRealm(consts.GOV_PATH)
+)
+
+/* HELPER */
```
Comment (jungju) [inline emission/__TEST_0_INIT_VARIABLE_AND_HELPER_test.gno:25]: Please adhere to Golang's comment conventions.
// gnsBalance ......

File: emission/distribution.gno [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+// checkSumDistributionPct ensures the sum of all distribution percentages is 100%
+func checkSumDistributionPct() {
+	sum := uint64(0)
+	for _, pct := range distributionPct {
+		sum += pct
+	}
+
+	if sum != 10000 {
+		panic(ufmt.Sprintf("sum of all pct should be 100%% (10000 bps), got %d\n", sum))
```
Comment (jungju) [inline emission/distribution.gno:152]: Panic automatically includes newlines.

**Review by jungju — CHANGES_REQUESTED** [review]

File: emission/distribution.gno (hunk 13 above)
Comment (r3v4s) [reply emission/distribution.gno:152]: AFAIK, gno panic doesn't include new lines.

@notJoon Could you confirm this?

File: emission/__TEST_0_INIT_VARIABLE_AND_HELPER_test.gno (hunk 12 above)
Comment (r3v4s) [reply emission/__TEST_0_INIT_VARIABLE_AND_HELPER_test.gno:25]: I don't think we need to write comments for test related functions

Comment (r3v4s) [reply emission/__TEST_0_INIT_VARIABLE_AND_HELPER_test.gno]: 33592c821dfd47060eed3af81baa6009e69a7db2

**Review by jungju — APPROVED** [review]

**Review by dongwon8247 — APPROVED** [review]

**Pull request merged**